src/test/scala/com/rbmhtechnology/calliope/GapDetectionSpec.scala

```diff
@@ -0,0 +1,225 @@
+  }
+
+  def waitForPersistTimeout(): Unit = {
+    sleep(timeout)
+  }
+
+  def filterGapLess[E: Sequenced](fromSeqNr: Long, events: immutable.Seq[E])(implicit gapDetection: GapDetection): immutable.Seq[E] =
+    events.filter(gapDetection.gapLess(fromSeqNr))
+
+  "A batch gap-detection" must {
+    "select all events if no gaps exist" in {
+      implicit val gaps = GapCollection.empty
+
+      filterGapLess(1, events(1 to 5)) mustBe events(1 to 5)
+    }
+    "select events up to a given gap" in {
+      implicit val g = gaps(4)
+
+      filterGapLess(1, events(1 to 5)) mustBe events(1 to 3)
+    }
+    "select events up to the first gap given consecutive gaps" in {
+      implicit val g = gaps(3, 4)
+
+      filterGapLess(1, events(1 to 5)) mustBe events(1 to 2)
+    }
+    "select no events if a gap at the beginning is detected" in {
+      implicit val g = gaps(1)
+
+      filterGapLess(1, events(1 to 5)) mustBe empty
+    }
+    "persist a gap after the persistence timeout" in {
+      implicit val g = gaps(6)
+
+      filterGapLess(1, events(1 to 10)) mustBe events(1 to 5)
+
+      waitForPersistTimeout()
+
+      filterGapLess(1, events(1 to 10)) mustBe events(1 to 10)
+    }
+    "persist a gap only if already detected" in {
+      implicit val g = gaps(6)
+
+      waitForPersistTimeout()
+
+      filterGapLess(1, events(1 to 10)) mustBe events(1 to 5)
+    }
+    "persist consecutive gaps after the persistence timeout" in {
+      implicit val g = gaps(6, 7, 8)
+
+      filterGapLess(1, events(1 to 10)) mustBe events(1 to 5)
+
+      waitForPersistTimeout()
+
+      filterGapLess(1, events(1 to 10)) mustBe events(1 to 10)
+    }
+    "select a missing event if it appears within the persistence timeout" in {
+      filterGapLess(1, events(1 to 5)(gaps(3))) mustBe events(1 to 2)
+
+      filterGapLess(3, events(3 to 5)) mustBe events(3 to 5)
+    }
+    "select a missing event within a consecutive gap if it appears within the persistence timeout" in {
+      filterGapLess(1, events(1 to 5)(gaps(3, 4))) mustBe events(1 to 2)
+
+      filterGapLess(3, events(3 to 5)(gaps(4))) mustBe events(3 to 3)
+
+      waitForPersistTimeout()
+
+      filterGapLess(4, events(4 to 5)(gaps(4))) mustBe events(5 to 5)
+    }
+    "select a missing event if it appears after the persistence timeout" in {
+      filterGapLess(1, events(1 to 5)(gaps(3))) mustBe events(1 to 2)
+
+      waitForPersistTimeout()
+
+      filterGapLess(3, events(3 to 5)) mustBe events(3 to 5)
+    }
+    "select all events after a previous after gaps were persisted" in {
+      implicit val g = gaps(4, 8)
+
+      filterGapLess(1, events(1 to 5)) mustBe events(1 to 3)
+
+      waitForPersistTimeout()
+
+      filterGapLess(4, events(4 to 9)) mustBe events(4 to 7)
+
+      waitForPersistTimeout()
+
+      filterGapLess(8, events(8 to 10)) mustBe events(8 to 10)
+
+      filterGapLess(1, events(1 to 5)) mustBe events(1 to 5)
```

Comment:
+1 should be a single check. 
This check is part of this test bc it should illustrate that once gaps have been persisted (in previous iterations) a subsequent iteration is not delayed by the (now persisted) gaps.